sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -2140,44 +2155,67 @@ void ProgramManager::removeImages(sycl_device_binaries DeviceBinary) {
     for (sycl_offload_entry EntriesIt = EntriesB; EntriesIt != EntriesE;
          EntriesIt = EntriesIt->Increment()) {
-
+      const char *Name = EntriesIt->GetName();
       // Drop entry for service kernel
-      if (std::strstr(EntriesIt->GetName(), "__sycl_service_kernel__")) {
-        m_ServiceKernels.erase(EntriesIt->GetName());
+      if (std::strstr(Name, "__sycl_service_kernel__")) {
+        removeFromMultimap(m_ServiceKernels, Name, Img);
         continue;
       }
 
       // Exported device functions won't have a kernel ID
-      if (m_ExportedSymbolImages.find(EntriesIt->GetName()) !=
-          m_ExportedSymbolImages.end()) {
+      if (m_ExportedSymbolImages.find(Name) != m_ExportedSymbolImages.end()) {
         continue;
       }
 
-      // remove everything associated with this KernelName
-      m_KernelUsesAssert.erase(EntriesIt->GetName());
-      m_KernelImplicitLocalArgPos.erase(EntriesIt->GetName());
+      // Remove everything associated with this KernelName if this is the last
+      // image referencing it, otherwise remove just the ID -> Img mapping.
+      auto RefCountIt = m_KernelNameRefCount.find(Name);
+      assert(RefCountIt != m_KernelNameRefCount.end());
+      int &RefCount = RefCountIt->second;
+      assert(RefCount > 0);
+      --RefCount;
 
-      if (auto It = m_KernelName2KernelIDs.find(EntriesIt->GetName());
+      if (auto It = m_KernelName2KernelIDs.find(Name);
           It != m_KernelName2KernelIDs.end()) {
-        m_KernelIDs2BinImage.erase(It->second);
-        m_KernelName2KernelIDs.erase(It);
+        if (RefCount == 0) {
+          m_KernelIDs2BinImage.erase(It->second);
+          m_KernelName2KernelIDs.erase(It);
+        } else {
+          removeFromMultimap(m_KernelIDs2BinImage, It->second, Img);
+        }
+      }
+
+      if (RefCount == 0) {
+        m_KernelUsesAssert.erase(Name);
+        m_KernelImplicitLocalArgPos.erase(Name);
+        m_KernelNameRefCount.erase(RefCountIt);
       }
     }
 
     // Drop reverse mapping
     m_BinImg2KernelIDs.erase(Img);
 
-    // Unregister exported symbols (needs to happen after the ID unmap loop)
+    // Unregister exported symbol -> Img pair (needs to happen after the ID
+    // unmap loop)
     for (const sycl_device_binary_property &ESProp :
          Img->getExportedSymbols()) {
-      m_ExportedSymbolImages.erase(ESProp->Name);
+      removeFromMultimap(m_ExportedSymbolImages, ESProp->Name, Img,
+                         /*AssertContains*/ false);
     }
 
     for (const sycl_device_binary_property &VFProp :
          Img->getVirtualFunctions()) {
       std::string StrValue = DeviceBinaryProperty(VFProp).asCString();
-      for (const auto &SetName : detail::split_string(StrValue, ','))
-        m_VFSet2BinImage.erase(SetName);
+      for (const auto &SetName : detail::split_string(StrValue, ',')) {
+        auto It = m_VFSet2BinImage.find(SetName);
+        assert(It != m_VFSet2BinImage.end());
```

Comment:
I'd suggest to create a tiny separate method for this.

Reply:
I'm not sure about outlining, but some inline comment would help...

Reply:
Added a comment and made some minor renaming changes to hopefully make this part more transparent.